eclipse/che-theia
# fix(workpace-plugin): Importing several private projects in a single devfile

### What does this PR do?
- Fix workspace start initial git cloning when multiple private projects are defined in the devfile.
- Only ask for credentials once if cloning from the same host.
- Adding a `devfile.yaml` for the workspace plugin (simplify review).
- ~Fix jest import errors in Che~
- Include unit tests

### What issues does this PR fix or reference?
Fix https://github.com/eclipse/che/issues/18494

### How to test this PR?
1. create 2 private projects in Github with README files
2. start a workspace from the given devfile.
3. build the plugin with the predifined tasks
4. run che-theia with the predefined task
5. go to the dashboard and add the private projects to the devfile of the running workspace. For instance (adapt the location with the ones you have created):
    ```
      - name: private-1
        source:
          location: 'https://github.com/sunix/test-private-1'
          type: github
      - name: private-2
        source:
          location: 'https://github.com/sunix/test-private-2'
          type: github
    ```
    SAVE the devfile DO NOT APPLY, DO NOT RESTART THE WORKSPACE!
6. back to che-theia, open the preview Url in a new tab, try to fill the username and password, observe the bug
7. stop the run task and close the preview tab, remove the content of `/tmp/theiadev_projects` from a terminal of che-theia-next-dev cotainer.
8. check the workspace devfile (dashboard) that the private projects are still there.
9. from the Github PR panel, checkout this PR
10. rebuild with the predifined task
11. rerun che-theia dev with the predifined task
12. open the Preview Url in a new tab.
13. fill your Github.com username and password, it should ask for it only once.
14. observe that the projects are cloned with the README files.


### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)


### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=next
